server/pages: Convert Jade to JSX templates

ockham opened the pull request:
Reboot of #4723 since the latest node (and possibly Babel) upgrades did miraculously solve some issues this branch had.

This PR uses `dangerouslySetInnerHtml` in `pages/index.jsx` to output the `renderToString`-rendered `Layout` element tree for SSR'd sections (=== Theme Sheets) as opposed to just putting that element tree into `pages/index.jsx`. This is done since
1. `express-react-views` (which this PR uses to render `pages/*.jsx`) internally uses `renderToStaticMarkup`, which loses  `data-reactid`s -- so its output can't be reconciled with the client generated React element tree.
2. We'd lose our current markup cache and would have to come up with a different solution.

Note that this PR removes the iOS specific magic from `index.jade` which has been unused for a while (see `10207-gh-calypso-pre-oss`).

Ideally, we'd also use the `EmptyContent` component for `404.jsx` and `500.jsx`, but that chokes on ES6 statements in that component for me.

Code QA:
- Make sure JSX files are up to par with their jade counterparts.

To test:
- Compare Page Source for this branch and `master` (or production), in particular:
  - For any logged-in section
  - For logged-out `/design`
  - For a logged-out 'Theme' (e.g. `/theme/twentyten`)
  - For a 404 (e.g. `/blah`)

Performance QA:
- Compare this branch to `master`

cc @gwwar @aduth @mtias @ehg for review

aduth commented on server/boot/index.js:
```diff
@@ -6,26 +6,35 @@ var path = require( 'path' ),
+	pages = require( 'pages' ),
+	ReactEngine = require( 'express-react-views' );
+
+var CALYPSO_ENV = process.env.CALYPSO_ENV || process.env.NODE_ENV || 'development';
```
1. `config( 'env' )` should handle the fallbacks on your behalf
2. I think we can/should start using Node-supported variable declarations (i.e. `const`)

ockham replied:
Heh, right. This PR's ancestor goes back a year to my early Calypso days 😊

aduth commented on server/pages/500.jsx:
```diff
@@ -0,0 +1,41 @@
+			<link rel="stylesheet" href={ styleCss } />
+		</head>
+		<body>
+			<div id="wpcom" className="wpcom-site">
```
Could we move some of the more repeated layout to a common component?

ockham replied:
I've opted to create a `Head` component to contain the HTML `head` stuff that's shared between our markup JSXes.

aduth commented on server/pages/desktop.jsx:
```diff
@@ -0,0 +1,96 @@
+
+class Desktop extends React.Component {
+	getStylesheet() {
+		const { env, isDebug, isRTL, urls } = this.props;
```
Camel-case variable naming applies to acronyms too: `isRTL` should be `isRtl` ([reference](https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/javascript.md#naming-conventions)).

aduth commented on server/pages/desktop.jsx:
```diff
@@ -0,0 +1,96 @@
+		return urls[ stylesheet ];
+	}
+
+	renderBadge() {
```
Should this be a separate / common component?

aduth commented on server/pages/desktop.jsx:
```diff
@@ -0,0 +1,96 @@
+		const { badge, feedbackURL } = this.props;
+		return (
+			<div className="environment-badge">
+				<a href={ feedbackURL } title="Report an issue" target="_blank" rel="noopener noreferrer" className="bug-report" />
```
Should we care to translate the `title`?

aduth commented on server/pages/desktop.jsx:
```diff
@@ -0,0 +1,96 @@
+					<div id="wpcom" className="wpcom-site">
+						<div className="wpcom-site__logo noticon noticon-wordpress" />
+					</div>
+					{ badge ? this.renderBadge() : null }
```
Hard to tell the expected type of `badge` because of lack of `propTypes`, but could this be `{ badge && this.renderBadge() }`?

ockham replied:
Yes, thanks! (I had it like that in one of the other JSXes).

aduth commented on server/pages/desktop.jsx:
```diff
@@ -0,0 +1,96 @@
+					{ i18nLocaleScript &&
+						<script src={ i18nLocaleScript } />
+					}
+					<script type="text/javascript" dangerouslySetInnerHTML={ { __html: // eslint-disable-line react/no-danger
```
- `type` should be unnecessary
- Curious: Does this need to be `dangerouslySetInnerHTML`?

ockham replied:
> Curious: Does this need to be `dangerouslySetInnerHTML`?

Seems to work without :-)

aduth commented on server/pages/index.jade:
```diff
@@ -1,135 +0,0 @@
-doctype html
-//
```
I will be sad to lose this 😢

ockham replied:
The doctype? `express-react-views` produces an HTML5 one by default, though, so we aren't really losing it.

mtias replied:
I think he meant the calypso ASCII art?

ockham replied:
Added it back. Was trickier than I hoped -- didn't manage to add an HTML comment, so it's now done via `dangerouslySetInnerHtml` in a `<script>` tag, contained in a `/*... */` comment.

aduth replied:
Another option is to pass it as part of `express-react-view` `doctype` option, since it's simply concatenated anyways:

https://github.com/reactjs/express-react-views/blob/c200d55/index.js#L50-L56

mtias commented on server/pages/404.jsx:
```diff
@@ -0,0 +1,43 @@
+		</head>
+		<body>
+			<div id="wpcom" className="wpcom-site">
+				<div className="layout has-no-sidebar">
```
Can we use `<Layout>` here or the sky would fall?

ockham replied:
`Layout` has too many non-SSR'able dependencies. Only `LoggedOutLayout` would work.

Also, see https://github.com/Automattic/wp-calypso/pull/8459#discussion_r81946504

mtias replied:
Can we make `Layout` be just a function that outputs the div with the classes (like Main and Sidebar?

mtias commented on server/pages/404.jsx:
```diff
@@ -0,0 +1,43 @@
+				<div className="layout has-no-sidebar">
+					<div id="content" className="layout__content">
+						<div id="primary" className="layout__primary">
+							<div className="empty-content">
```
Can we use `<EmptyContent>` here?

ockham replied:
I wish!

> Ideally, we'd also use the `EmptyContent` component for `404.jsx` and `500.jsx`, but that chokes on ES6 statements in that component for me.

(from the PR desc ;-)

I hope that we'll eventually be able -- it's part of this PR's point.

ockham replied:
From https://github.com/reactjs/express-react-views#views:

> Under the hood, [Babel][babel] is used to compile your views into ES5 friendly code, using the [react][babel-preset-react] and [es2015][babel-preset-es2015] presets by default. **Only the files in your `views` directory (i.e. `app.set('views', __dirname + '/views')`) will be compiled.**

(Highlight mine)

ockham commented:
Okay, I think I've addressed all feedback.

As for componentization, no specific `Shell` or `Layout` components yet; rather, I've chosen to factor out `head` into a shared `Head` component, which I think goes a long way for de-duping and re-use already. I think most of `head` is boilerplate and should be the same for all our server-side templates (think favicons in different sizes) -- they seem to have gotten out of sync quite badly.

ockham commented:
Timing-wise, looking at server webpack build and `wp-calypso booted in` times, `master` and this branch are pretty much tied.

Anyone up for giving this another look (or a :+1: :smile: )?

mtias commented on server/pages/badge.jsx:
```diff
@@ -0,0 +1,47 @@
+);
+
+const DevDocsLink = ( { devDocsUrl, docs } ) => (
+  <span className="environment is-docs">
```
Seems to have spaces instead of tabs.

ockham replied:
Why u fail me, Atom?

aduth replied:
Now to be caught by pre-commit hooks and CircleCI 😄 

https://github.com/Automattic/wp-calypso/pull/8565

mtias commented on server/pages/index.jsx:
```diff
@@ -0,0 +1,115 @@
+import Head from './head';
+import Badge from './badge';
+
+class Markup extends React.Component {
```
Maybe call this `Document`?

aduth commented on server/pages/badge.jsx:
```diff
@@ -0,0 +1,47 @@
+	</span>
+);
+
+const DevDocsLink = ( { devDocsUrl, docs } ) => (
```
The docs link seems to have disappeared:

![image](https://cloud.githubusercontent.com/assets/1779930/19192611/c87c41ea-8c74-11e6-9465-c8779fe9c56a.png)

(Should be to the right of branch name)

ockham commented on server/pages/head.jsx:
```diff
@@ -0,0 +1,48 @@
+/**
```
Moved `head.jsx` and `badge.jsx` to a `components/` subdir.

ockham commented:
Okay, addressed all feedback.

aduth commented on server/pages/components/badge.jsx:
```diff
@@ -0,0 +1,46 @@
+/**
```
Could these live in `client/components` ?

aduth commented:
>Ideally, we'd also use the `EmptyContent` component for `404.jsx` and `500.jsx`, but that chokes on ES6 statements in that component for me.

Wonder why this is? Do we need to be concerned about ES6 transpilation?

ockham commented:
>>Ideally, we'd also use the `EmptyContent` component for `404.jsx` and `500.jsx`, but that chokes on ES6 statements in that component for me.

> Wonder why this is? Do we need to be concerned about ES6 transpilation?

See https://github.com/Automattic/wp-calypso/pull/8459#discussion_r81956801

aduth commented:
>See #8459 (comment)

That's kinda gross. It even uses its own separate Babel configuration. Looking more at the code for `express-react-views`, I lean more toward rolling our own instead.

ockham commented:
Seems like it's possible to turn that off, though: https://github.com/reactjs/express-react-views#options

Makes me wonder though why it doesn't pick up our pre-babelled code anyway (outside the `views` dir, I mean).

ockham commented:
But yeah, their implementation is pretty small in terms of LOCs, so I guess we could do it ourselves.

gwwar commented:
We should also test this for Desktop envs before shipping this. cc @sendhil for a heads up

ockham commented:
Okay, the new home-grown rendering engine should allow us to import other components, and even move the JSX templates inside `server/pages` to `client/some-subdir`, if we're so inclined.

aduth commented on server/pages/components/head.jsx:
```diff
@@ -0,0 +1,38 @@
+import React from 'react';
+
+const Head = ( { children, stylesheetUrl, title = 'WordPress.com', faviconUrl = '//s1.wp.com/i/favicon.ico' } ) => (
+  <head>
```
Spaces should be tabs.

aduth commented on server/boot/index.js:
```diff
@@ -1,20 +1,52 @@
+import config from 'config';
+import pages from 'pages';
+
+function renderJsxFile( filename, options, callback ) {
```
What about HTML minification for production?

ockham replied:
Is it worth it? `renderToStaticMarkup` doesn't seem to add any whitespace in between HTML tags, so I wonder if the few bytes we might be able to shave off are worth the overhead of running it thru a minifier...

aduth replied:
Oh, I didn't realize that the output is already minified. In that case, we're golden!

ockham commented:
> I think the fragment components we've created in `/pages` may be better off in `/client/components`. What do others think?

I think that makes sense. I'd go so far as to also move `index.jsx`, `desktop.jsx`, `404.jsx`, and `500.jsx` somewhere below `client/` (`client/document/` maybe?) to have all JSXes in one place.

mtias commented:
We should rename `client` for that to make the most sense :)

mtias commented:
> I think the fragment components we've created in /pages may be better off in /client/components

I think some of them may belong in `client/layout`, and the others could be in `client/document`. It'd be good to think about how to easily extend / overwrite these for a project.

ockham commented:
> We should rename `client` for that to make the most sense :)

There's an ~~app~~ [issue](https://github.com/Automattic/wp-calypso/issues/2319) for that :-)

ockham commented:
I moved the "document" JSXes to `client/document`, and `badge` and `head` to `client/components`, where I think right now they fit best. To the latter two, I also added `README`s and tests. While working on those, I realized some more refactoring was in order to encapsulate them better. (Now that they are components, we can assign some variables at component or at least module level -- no need to do it in `server/pages`.)

This could be pushed further, e.g. by `connect()`ing `client/document` (and removing some code from `server/render`). However, I feel PR/refactoring fatigue creeping in, so I think we should keep the cherries and icing for subsequent PRs.

Kindly give this PR another review then, pls :-)

stephanethomas commented on client/components/badge/index.jsx:
```diff
@@ -0,0 +1,114 @@
+			{ branchName }
+		</span>
+	);
+};
```
I suspect the `<Branch />` component is also rendered client-side at some point, right? If that's the case, shouldn't we avoid calling `getCurrentBranchName()` and `getCurrentCommitShortChecksum()` on each render, since they are costly i/o operations?

ockham replied:
> I suspect the `<Branch />` component is also rendered client-side at some point, right?

Not really -- this component has basically been factored out of the converted templates, so they're only ever rendered on the server side. I'll add a note to the component's `README`.

> If that's the case, shouldn't we avoid calling `getCurrentBranchName()` and `getCurrentCommitShortChecksum()` on each render, since they are costly i/o operations?

I suspect they won't work at all on the client, since they rely on node's `child_process` module (and on `git`). But they're also only used in the `development` environment.

ehg commented:
Are we going to hit any performance concerns here? Does `express` cache the view templates for us, so we don't `renderToStaticMarkup` on every request?

ehg commented on server/pages/README.md:
```diff
@@ -3,4 +3,5 @@ Pages
 
 This module renders the main HTML document and handles appending a query string to `build.js` and `style.css` that changes when those files are modified.
-Like Cloudup, we are using [jade](http://jade-lang.com/) as the templating language, although at some point, we may switch to using React server-side.
+We use [`express-react-views`](https://github.com/reactjs/express-react-views)
```
Not anymore :)

ockham commented:
> Are we going to hit any performance concerns here? Does `express` cache the view templates for us, so we don't `renderToStaticMarkup` on every request?

I'm afraid it doesn't (it only caches the view files, i.e. prior to rendering). Think we should cache? Could be a bit complex (especially dealing with the other cache in `server/render`). FWIW, we didn't have caching with Jade before...

ockham commented:
Stalled right now while I'm busy working on https://github.com/Automattic/wp-calypso/milestone/136. I'll get back to afterwards; I might end up moving the caching bits from `server/render` to `server/pages` to also cache the JSX generated there.

alisterscott commented:
Closing as this PR is stale - please reopen if necessary

ockham commented:
Might be worth revisiting after the React 16 upgrade.
